vm/src/pyobject.rs

```diff
@@ -700,16 +700,23 @@ where
 }
 
 impl PyObject<dyn PyObjectPayload> {
-    pub fn downcast<T: PyObjectPayload>(self: Rc<Self>) -> Option<PyRef<T>> {
+    /// Attempt to downcast this reference to a subclass.
+    ///
+    /// If the downcast fails, the original ref is returned in as `Err` so
+    /// another downcast can be attempted without unnecessary cloning.
+    ///
+    /// Note: The returned `Result` is _not_ a `PyResult`, even though the
+    ///       types are compatible.
+    pub fn downcast<T: PyObjectPayload>(self: Rc<Self>) -> Result<PyRef<T>, PyObjectRef> {
```

Comment:
I thought `arbitrary self types` hadn't landed yet?

Reply:
It hasn't for user-defined types, but `Box<Self>` and `Rc<Self>` as self types are allowed.